### PR TITLE
Make waveform generator a NMI to run always, increase accuracy

### LINKED ISSUE #5568: PWM is blocked during SPIFFS write operations
### Basic Infos

- [X] This issue complies with the [issue POLICY doc](https://github.com/esp8266/Arduino/blob/master/POLICY.md).
- [X] I have read the documentation at [readthedocs](https://arduino-esp8266.readthedocs.io/en/latest) and the issue is not addressed there.
- [X] I have tested that the issue is present in current master branch (aka latest git).
- [X] I have searched the issue tracker for a similar issue.
- [X] If there is a stack dump, I have decoded it.  (no stack dump)
- [X] I have filled out all fields below.

#### Platform

- Hardware:					ESP-12
- Core Version:			b26c19e82e1b6c94f86a66dc2652336e8ca5bba3
- Development Env:		Platformio / Arduino IDE
- Operating System:		Windows

### Settings in IDE

- Module:  			Wemos D1 mini
(removed rest because it's not a flashing problem)

### Problem Description

The ESP8266 does not have hardware support for PWM, so the Arduino Core implements PWM in software with timer interrupts.

However, these PWM interrupts seem to be blocked by *some* SPIFFS write operations. In the picture below, every yellow spike shows the start of a small write operation. The green channel shows a PWM at 500 Hz.

As you can see, every 16th or so write operation takes a lot longer than the other ones (about 110 ms), and disturbs the PWM output.

[![oscilloscope image][1]][1]

I understand that SPIFFS is not realtime capable and it is unavoidable that some write operations take longer than others. But maybe the SPIFFS code can be changed so that the PWM is not affected by that?

  [1]: https://i.stack.imgur.com/MIlET.png

### [MCVE](https://stackoverflow.com/help/mcve) Sketch

```cpp
#include <Arduino.h>
#include <FS.h>

void setup() {
    pinMode(14, OUTPUT);

    analogWriteRange(255);
    analogWriteFreq(500);
    analogWrite(5, 128);   // GPIO5 (pin "D1" on D1 mini)

    SPIFFS.begin();
}

String toWrite = "Hello world!";

void loop() {
    // Trigger for oscilloscope (yellow channel):
    digitalWrite(14, 1);
    delayMicroseconds(100);
    digitalWrite(14, 0);

    File f = SPIFFS.open("/test.txt", "w");
    f.write((uint8_t*)toWrite.c_str(), toWrite.length());
    f.close();
}

```

### Debug Messages

(see image)

### PR DESCRIPTION
**Pull request by earlephilhower**
Make the waveform generator an NMI using the same code as in 2.4.0.
Making it NMI will ensure it runs even when interrupts are disabled.

Fixes #5568

### CODE REVIEW

**Comment by earlephilhower** [comment]
Ran the SpeedTest.ino on SPIFFS in PR #5511 with a 1KHz PWM output on a single pin and it's looking good.

Attempted the WiFi Server test at https://gist.github.com/arunoda/3f17eba4f3f1fd5e1a7adc86e7b62ca7 with a 1KHz output on a single pin as well, TCP v2 hi BW, 160MHz, and got about 670KByte/sec (5.3Mbit/sec) downloads from the 8266.

At 10KHz I can get occasional SPIFFS WDT timeouts when writing 512KByte at a time in 256byte chunks (but adding in yields() every 32KB fixes this while reducing write BW by under 10%, so it's not really a showstopper), but WiFi works fine and delivers ~620KBytes/sec (4.96Kbit/sec) doe to increased CPU usage by the PWM generator.

**Comment by earlephilhower** [comment]
Argh, looking over it one last time I realize it's now going to be a problem as-adjusted.

Before we could twiddle with IRQ data structures as long as we bracketed it with disable-irq/.../enable-irq.

For NMI, disable/enable irq doesn't do anything.  So we could catch an interrupt while in the middle of changing values and bad stuff will happen.

Need to make it lockless somehow, may be a significant effort.

File: cores/esp8266/core_esp8266_waveform.c [hunk 1]
```diff
@@ -38,84 +38,41 @@
...
-// Set/clear *any* GPIO
-#define SetGPIOPin(a) do { if (a < 16) { GPOS |= (1<<a); } else { GP16O |= 1; } } while (0)
-#define ClearGPIOPin(a) do { if (a < 16) { GPOC |= (1<<a); } else { GP16O &= ~1; } } while (0)
-// Set/clear GPIO 0-15
+// Set/clear GPIO 0-15 by bitmask
 #define SetGPIO(a) do { GPOS = a; } while (0)
 #define ClearGPIO(a) do { GPOC = a; } while (0)
 
+// Convert from us to ESP8266 clocks
+#define MicrosecondsToCycles(m) (clockCyclesPerMicrosecond() * (m))
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: This is already defined in Arduino.h as  `clockCyclesPerMicrosecond`

File: cores/esp8266/core_esp8266_waveform.c [hunk 2]
```diff
@@ -38,84 +38,41 @@
...
-  {0, 0, 1<<10, 0, 0, 0, 0, 0},
-#endif
-  // GPIO 11 not allowed, used for flash
-  {0, 0, 1<<12, 0, 0, 0, 0, 0},
-  {0, 0, 1<<13, 0, 0, 0, 0, 0},
-  {0, 0, 1<<14, 0, 0, 0, 0, 0},
-  {0, 0, 1<<15, 0, 0, 0, 0, 0},
-  {0, 0, 0, 1, 0, 0, 0, 0}  // GPIO16
-};
+static Waveform waveform[17];        // State of all possible pins
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: The isFlashInterfacePin(9) macro no longer applies here?

File: cores/esp8266/core_esp8266_waveform.c [hunk 3]
```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
...
+  uint32_t timeoutCycle = GetCycleCountIRQ() + MicrosecondsToCycles(14);
+
+  if (waveformToEnable || waveformToDisable) {
+    // Handle enable/disable requests from main app.
+    waveformEnabled = (waveformEnabled & ~waveformToDisable) | waveformToEnable; // Set the requested waveforms on/off
+    waveformState &= ~waveformToEnable;  // And clear the state of any just started
+    waveformToEnable = 0;
+    waveformToDisable = 0;
+    startPin = __builtin_ffs(waveformEnabled) - 1;
+    endPin = 32 - __builtin_clz(waveformEnabled);
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: I don't understand these two lines. Maybe add comments?

File: cores/esp8266/core_esp8266_waveform.c [hunk 4]
```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
...
+      Waveform *wave = &waveform[i];
+      uint32_t now = GetCycleCountIRQ();
+
+      // Disable any waveforms that are done
+      if (wave->expiryCycle) {
+        int32_t expiryToGo = wave->expiryCycle - now;
+        if (expiryToGo < 0) {
+            // Done, remove!
+            waveformEnabled &= ~mask;
+            if (i==16) GP16O &= ~1;
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: if-else astyle
```
if(cond)
  blah
else
  bleh
```

File: cores/esp8266/core_esp8266_waveform.c [hunk 5]
```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
...
+          else SetGPIO(mask);
+
           wave->nextServiceCycle = now + wave->nextTimeHighCycles;
           nextEventCycles = min_u32(nextEventCycles, wave->nextTimeHighCycles);
         } else {
-          ClearGPIO(wave->gpioMask);
-          if (wave->gpio16Mask) {
-            GP16O &= ~wave->gpio16Mask;
-          }
+          if (i==16) GP16O &= ~1; // GPIO16 write slow as it's RMW
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: if-else astyle

File: cores/esp8266/core_esp8266_waveform.c [hunk 6]
```diff
@@ -273,47 +259,27 @@ static ICACHE_RAM_ATTR void timer1Interrupt() {
...
-      } else {
-        uint32_t newTimeLeftCycles = wave->timeLeftCycles - deltaCycles;
-        wave->timeLeftCycles = newTimeLeftCycles;
-      }
-    }
-  }
+    uint32_t now = GetCycleCountIRQ();
+    done = false;
+    int32_t cycleDeltaNextEvent = timeoutCycle - (now + nextEventCycles);
+    if (cycleDeltaNextEvent < 0) done = true;
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: No single line ifs please

File: cores/esp8266/core_esp8266_waveform.c [hunk 7]
```diff
@@ -273,47 +259,27 @@ static ICACHE_RAM_ATTR void timer1Interrupt() {
...
+    int32_t cyclesLeftTimeout = timeoutCycle - now;
+    if (cyclesLeftTimeout < 0) done = true;
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: No single line ifs pls

File: cores/esp8266/core_esp8266_waveform.c [hunk 8]
```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
...
       if (cyclesToGo < 0) {
-        wave->state = !wave->state;
-        if (wave->state) {
-          SetGPIO(wave->gpioMask);
-          if (wave->gpio16Mask) {
-            GP16O |= wave->gpio16Mask; // GPIO16 write slow as it's RMW
-          }
+        waveformState ^= mask;
+        if (waveformState & mask) {
+          if (i==16) GP16O |= 1; // GPIO16 write slow as it's RMW
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: if-else astyle

File: cores/esp8266/core_esp8266_waveform.c [hunk 9]
```diff
@@ -149,122 +104,153 @@ void setTimer1Callback(uint32_t (*fn)()) {
...
+  if (!(waveformEnabled & mask)) {
+    return false; //It's not running, nothing to do here
   }
-  return false;
+  waveformToDisable |= mask;
+  // Ensure tiomely service....
+  if (T1L > MicrosecondsToCycles(10)) {
+    timer1_write(MicrosecondsToCycles(10));
+  }
+  while (waveformToDisable) {
```
Comment (devyte) [inline cores/esp8266/core_esp8266_waveform.c]: Does it make sense to wait with timeout?

Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: Good eyes, but at this point we know the timer is running and it has, at most, 10us of runtime before the NMI will be called.  So no need to have a timeout here.

File: cores/esp8266/core_esp8266_waveform.c (hunk 3 above)
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: I'm happily surprised to stump you. :) They're GCC's Find-First-Set (plus one) and Count-Leading-Zeros.  I'll add a comment.

File: cores/esp8266/core_esp8266_waveform.c (hunk 2 above)
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: No, that's taken care of in `startWaveform` where we return `false` if it's an invalid pin or `isFlashInterfacePin()`.  The simpler array makes for simpler (faster) code in the IRQ.

File: cores/esp8266/core_esp8266_waveform.c (hunk 1 above)
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: Yes, ` clockCyclesPerMicrosecond` is defined, but I need it for a fixed # of milliseconds (m) bit.

Comment (devyte) [reply cores/esp8266/core_esp8266_waveform.c]: I don't follow, the Arduino.h macro looks the same as this one:
```
#define microsecondsToClockCycles(a) ( (a) * clockCyclesPerMicrosecond() )
```

Also: milliseconds? I'm confused now.

Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: Ah, darn it.  I did mean *micro*seconds and I didn't see that macro for some reason.  I'll replace the lot out...

File: cores/esp8266/core_esp8266_waveform.c [hunk 10]
```diff
@@ -53,70 +50,35 @@
...
 
-static inline ICACHE_RAM_ATTR uint32_t min_u32(uint32_t a, uint32_t b) {
-  if (a < b) {
-    return a;
-  }
-  return b;
-}
+// Enable lock-free by only allowing updates to waveformState and waveformEnabled from IRQ service routine
+static volatile uint32_t waveformToEnable = 0; // Message from startWaveform to IRQ to actuall begin a wvf
+static volatile uint32_t waveformToDisable = 0; // Message from startWaveform to IRQ to actuall begin a wvf
```
[the remark this answers is not in the document]
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: This and other commenting/if{}else{} bracketing taken care of in latest push

File: cores/esp8266/core_esp8266_waveform.c (hunk 1 above)
Comment (earlephilhower) [reply cores/esp8266/core_esp8266_waveform.c]: Done in latest push.

**Comment by Jason2866** [comment]
With this change Tasmota is nearly unresponsible. Before it worked very well.

**Comment by earlephilhower** [comment]
@Jason2866, please open an issue and be much more specific or we're not going to be able to help you out.  Also, are you sure it's related to this specific PR and not #5513 which was noticed at the 1st 2.5.0-dev release.   Thx, -EFP3

**Comment by Jason2866** [comment]
@earlephilhower Thx for your comment. Did more testing. At the end the behaviour was a part defective Wemos Mini!! Everything is fine now :-)

**Comment by davidmpye** [comment]
This is causing me a problem :-(

If you try to use something timing critical like FastLED (or Adafruit_Neopixel) *and* PWM at the same time, you run into issues.

When I write the LED data out, I need to have interrupts disabled, otherwise, the NMI firing causes the FastLED data sequence to glitch, making the LEDs flash randomly.    The fastled library suggests disabling interrupts during the write sequence, but obviously this fails if using NMI.

Is it possible to create a #define or similar to set whether the PWM code uses NMI or not?

David

**Comment by earlephilhower** [comment]
@davidmpye the most compatible is going to be stopWaveform/etc. right before the code in question followed by startWaveform/etc. right after.  There may be other code in the processor (WiFi, etc.) which also comes in the binary SDK.  This isn't as bad as it sounds since the tone/analogWrite would be garbage while writing the LEDs anyway since w/IRQs disabled they won't do anything at all and will be stuck at full high or low.

Alternately, look in the core_esp8266_waveform.cpp file where it's pretty straightforward to hack code in to make it a normal IRQ (but will fail during SPIFFS writes or other periods when IRQs are disabled for any period).

**Pull request merged**